File: agent_matchmaking/pricing.py
"""Price discovery mechanisms — posted price, RFQ, and auction (Section 8).

Supports three mechanisms selectable per match request.
"""

import time as _time
from typing import Any, Callable, Dict, List, Optional, Tuple

from .schema import (
    MatchRequest,
    UnifiedCapabilityProfile,
    _gen_id,
    _now_iso,
)


# ---------------------------------------------------------------------------
# Price estimation
# ---------------------------------------------------------------------------

def estimate_cost(
    ucp: UnifiedCapabilityProfile,
    task_complexity: float = 1.0,
) -> float:
    """Estimate cost for a task based on UCP pricing model.

    Args:
        ucp: The agent's UCP with cost section.
        task_complexity: Multiplier for variable costs (1.0 = average).

    Returns:
        Estimated total cost in the agent's base currency.
    """
    base = ucp.cost.base_rate.amount
    variable = ucp.cost.variable_rate.amount * task_complexity * 1000  # assume ~1000 units
    return base + variable


# ---------------------------------------------------------------------------
# Mechanism 1: Posted Price (Section 8.2)
# ---------------------------------------------------------------------------

class PostedPriceResult:
    """Result from posted-price mechanism."""

    __slots__ = ("agent_id", "estimated_cost", "currency", "pricing_model")

    def __init__(
        self,
        agent_id: str = "",
        estimated_cost: float = 0.0,
        currency: str = "USD",
        pricing_model: str = "posted_price",
    ) -> None:
        self.agent_id = agent_id
        self.estimated_cost = estimated_cost
        self.currency = currency
        self.pricing_model = pricing_model

    def to_dict(self) -> Dict[str, Any]:
        return {
            "agent_id": self.agent_id,
            "estimated_cost": round(self.estimated_cost, 4),
            "currency": self.currency,
            "pricing_model": self.pricing_model,
        }


def posted_price(
    candidates: List[UnifiedCapabilityProfile],
    task_complexity: float = 1.0,
) -> List[PostedPriceResult]:
    """Mechanism 1: Estimate costs from posted pricing models."""
    results = []
    for ucp in candidates:
        cost = estimate_cost(ucp, task_complexity)
        results.append(PostedPriceResult(
            agent_id=ucp.identity.amp_id,
            estimated_cost=cost,
            currency=ucp.cost.base_rate.currency,
            pricing_model=ucp.cost.pricing_model,
        ))
    return results


# ---------------------------------------------------------------------------
# Mechanism 2: Request for Quote (Section 8.2)
# ---------------------------------------------------------------------------

class Quote:
    """A price quote from an agent."""

    __slots__ = (
        "quote_id", "agent_id", "amount", "currency",
        "terms", "validity_window_ms", "timestamp",
    )

    def __init__(
        self,
        agent_id: str = "",
        amount: float = 0.0,
        currency: str = "USD",
        terms: str = "",
        validity_window_ms: int = 3600000,
        quote_id: str = "",
        timestamp: str = "",
    ) -> None:
        self.quote_id = quote_id or _gen_id("quote")
        self.agent_id = agent_id
        self.amount = amount
        self.currency = currency
        self.terms = terms
        self.validity_window_ms = validity_window_ms
        self.timestamp = timestamp or _now_iso()

    def to_dict(self) -> Dict[str, Any]:
        return {
            "quote_id": self.quote_id,
            "agent_id": self.agent_id,
            "amount": round(self.amount, 4),
            "currency": self.currency,
            "terms": self.terms,
            "validity_window_ms": self.validity_window_ms,
            "timestamp": self.timestamp,
        }

    @classmethod
    def from_dict(cls, d: Dict[str, Any]) -> "Quote":
        return cls(
            quote_id=d.get("quote_id", ""),
            agent_id=d.get("agent_id", ""),
            amount=d.get("amount", 0.0),
            currency=d.get("currency", "USD"),
            terms=d.get("terms", ""),
            validity_window_ms=d.get("validity_window_ms", 3600000),
            timestamp=d.get("timestamp", ""),
        )


class RFQSession:
    """Manages a Request for Quote session (Section 8.2).

    Flow:
        1. Create RFQ with task description
        2. Solicit quotes from matched agents
        3. Collect quotes
        4. Select best quote
    """

    def __init__(self, request: MatchRequest) -> None:
        self.rfq_id = _gen_id("rfq")
        self.request = request
        self.quotes: List[Quote] = []
        self.selected_quote: Optional[Quote] = None
        self.created_at = _now_iso()

    def add_quote(self, quote: Quote) -> None:
        self.quotes.append(quote)

    def rank_quotes(
        self,
        trust_scores: Optional[Dict[str, float]] = None,
        quality_weight: float = 0.5,
        price_weight: float = 0.5,
    ) -> List[Quote]:
        """Rank quotes by combined price + trust score.

        Lower price = better. Higher trust = better.
        """
        trust_scores = trust_scores or {}

        scored: List[Tuple[float, Quote]] = []
        if not self.quotes:
            return []

        max_amount = max(q.amount for q in self.quotes) or 1.0

        for q in self.quotes:
            # Price score: lower is better, normalized to [0, 100]
            price_score = (1.0 - q.amount / max_amount) * 100 if max_amount > 0 else 50.0
            trust = trust_scores.get(q.agent_id, 50.0)
            combined = price_weight * price_score + quality_weight * trust
            scored.append((combined, q))

        scored.sort(key=lambda x: -x[0])
        return [q for _, q in scored]

    def select(self, quote_id: str) -> Optional[Quote]:
        """Accept a quote by ID."""
        for q in self.quotes:
            if q.quote_id == quote_id:
                self.selected_quote = q
                return q
        return None

    def to_dict(self) -> Dict[str, Any]:
        return {
            "rfq_id": self.rfq_id,
            "request_id": self.request.request_id,
            "quotes": [q.to_dict() for q in self.quotes],
            "selected_quote": self.selected_quote.to_dict() if self.selected_quote else None,
            "created_at": self.created_at,
        }


# ---------------------------------------------------------------------------
# Mechanism 3: Auction — Vickrey sealed-bid second-price (Section 6.3/8.2)
# ---------------------------------------------------------------------------

class Bid:
    """A bid in an auction."""

    __slots__ = ("bid_id", "agent_id", "amount", "currency", "timestamp")

    def __init__(
        self,
        agent_id: str = "",
        amount: float = 0.0,
        currency: str = "USD",
        bid_id: str = "",
        timestamp: str = "",
    ) -> None:
        self.bid_id = bid_id or _gen_id("bid")
        self.agent_id = agent_id
        self.amount = amount
        self.currency = currency
        self.timestamp = timestamp or _now_iso()

    def to_dict(self) -> Dict[str, Any]:
        return {
            "bid_id": self.bid_id,
            "agent_id": self.agent_id,
            "amount": round(self.amount, 4),
            "currency": self.currency,
            "timestamp": self.timestamp,
        }

    @classmethod
    def from_dict(cls, d: Dict[str, Any]) -> "Bid":
        return cls(
            bid_id=d.get("bid_id", ""),
            agent_id=d.get("agent_id", ""),
            amount=d.get("amount", 0.0),
            currency=d.get("currency", "USD"),
            timestamp=d.get("timestamp", ""),
        )


class AuctionResult:
    """Result of an auction."""

    __slots__ = (
        "auction_id", "format", "winner_id",
        "winning_bid", "clearing_price",
        "total_bids", "timestamp",
    )

    def __init__(
        self,
        auction_id: str = "",
        auction_format: str = "vickrey",
        winner_id: str = "",
        winning_bid: float = 0.0,
        clearing_price: float = 0.0,
        total_bids: int = 0,
        timestamp: str = "",
    ) -> None:
        self.auction_id = auction_id or _gen_id("auction")
        self.format = auction_format
        self.winner_id = winner_id
        self.winning_bid = winning_bid
        self.clearing_price = clearing_price
        self.total_bids = total_bids
        self.timestamp = timestamp or _now_iso()

    def to_dict(self) -> Dict[str, Any]:
        return {
            "auction_id": self.auction_id,
            "format": self.format,
            "winner_id": self.winner_id,
            "winning_bid": round(self.winning_bid, 4),
            "clearing_price": round(self.clearing_price, 4),
            "total_bids": self.total_bids,
            "timestamp": self.timestamp,
        }


def vickrey_auction(
    bids: List[Bid],
    min_trust_score: float = 0.0,
    trust_scores: Optional[Dict[str, float]] = None,
) -> Optional[AuctionResult]:
    """Sealed-bid second-price (Vickrey) auction (Section 6.3).

    Lowest bidder wins, pays the second-lowest price.
    Bidders below min_trust_score are disqualified.
    """
    trust_scores = trust_scores or {}

    # Filter by trust threshold
    qualified = [
        b for b in bids
        if trust_scores.get(b.agent_id, 100.0) >= min_trust_score
    ]

    if not qualified:
        return None

    # Sort by amount (ascending — lowest wins)
    qualified.sort(key=lambda b: b.amount)

    winner = qualified[0]
    # Second-price: winner pays the second-lowest bid
    clearing_price = qualified[1].amount if len(qualified) > 1 else winner.amount

    return AuctionResult(
        auction_format="vickrey",
        winner_id=winner.agent_id,
        winning_bid=winner.amount,
        clearing_price=clearing_price,
        total_bids=len(bids),
    )


def english_auction(
    bids: List[Bid],
    min_trust_score: float = 0.0,
    trust_scores: Optional[Dict[str, float]] = None,
) -> Optional[AuctionResult]:
    """English (descending-price) auction.

    For agent matching, "English" means agents bid progressively lower.
    Last remaining bidder (lowest price meeting trust threshold) wins.
    """
    trust_scores = trust_scores or {}

    qualified = [
        b for b in bids
        if trust_scores.get(b.agent_id, 100.0) >= min_trust_score
    ]

    if not qualified:
        return None

    qualified.sort(key=lambda b: b.amount)
    winner = qualified[0]

    return AuctionResult(
        auction_format="english",
        winner_id=winner.agent_id,
        winning_bid=winner.amount,
        clearing_price=winner.amount,
        total_bids=len(bids),
    )

File: typescript/src/pricing.ts
import {
  MatchRequest,
  UnifiedCapabilityProfile,
  genId,
  nowIso,
} from "./schema";

// ---------------------------------------------------------------------------
// Price estimation
// ---------------------------------------------------------------------------

export function estimateCost(ucp: UnifiedCapabilityProfile, taskComplexity = 1.0): number {
  const base = ucp.cost.baseRate.amount;
  const variable = ucp.cost.variableRate.amount * taskComplexity * 1000;
  return base + variable;
}

// ---------------------------------------------------------------------------
// Mechanism 1: Posted Price (Section 8.2)
// ---------------------------------------------------------------------------

export interface PostedPriceResultDict {
  agent_id: string;
  estimated_cost: number;
  currency: string;
  pricing_model: string;
}

export class PostedPriceResult {
  agentId: string;
  estimatedCost: number;
  currency: string;
  pricingModel: string;

  constructor(agentId = "", estimatedCost = 0, currency = "USD", pricingModel = "posted_price") {
    this.agentId = agentId;
    this.estimatedCost = estimatedCost;
    this.currency = currency;
    this.pricingModel = pricingModel;
  }

  toDict(): PostedPriceResultDict {
    return {
      agent_id: this.agentId,
      estimated_cost: Math.round(this.estimatedCost * 10000) / 10000,
      currency: this.currency,
      pricing_model: this.pricingModel,
    };
  }
}

export function postedPrice(
  candidates: UnifiedCapabilityProfile[],
  taskComplexity = 1.0,
): PostedPriceResult[] {
  return candidates.map((ucp) => new PostedPriceResult(
    ucp.identity.ampId,
    estimateCost(ucp, taskComplexity),
    ucp.cost.baseRate.currency,
    ucp.cost.pricingModel,
  ));
}

// ---------------------------------------------------------------------------
// Mechanism 2: Request for Quote (Section 8.2)
// ---------------------------------------------------------------------------

export interface QuoteDict {
  quote_id: string;
  agent_id: string;
  amount: number;
  currency: string;
  terms: string;
  validity_window_ms: number;
  timestamp: string;
}

export class Quote {
  quoteId: string;
  agentId: string;
  amount: number;
  currency: string;
  terms: string;
  validityWindowMs: number;
  timestamp: string;

  constructor(opts: {
    quoteId?: string;
    agentId?: string;
    amount?: number;
    currency?: string;
    terms?: string;
    validityWindowMs?: number;
    timestamp?: string;
  } = {}) {
    this.quoteId = opts.quoteId || genId("quote");
    this.agentId = opts.agentId ?? "";
    this.amount = opts.amount ?? 0;
    this.currency = opts.currency ?? "USD";
    this.terms = opts.terms ?? "";
    this.validityWindowMs = opts.validityWindowMs ?? 3600000;
    this.timestamp = opts.timestamp || nowIso();
  }

  toDict(): QuoteDict {
    return {
      quote_id: this.quoteId,
      agent_id: this.agentId,
      amount: Math.round(this.amount * 10000) / 10000,
      currency: this.currency,
      terms: this.terms,
      validity_window_ms: this.validityWindowMs,
      timestamp: this.timestamp,
    };
  }

  static fromDict(d: Partial<QuoteDict>): Quote {
    return new Quote({
      quoteId: d.quote_id ?? "",
      agentId: d.agent_id ?? "",
      amount: d.amount ?? 0,
      currency: d.currency ?? "USD",
      terms: d.terms ?? "",
      validityWindowMs: d.validity_window_ms ?? 3600000,
      timestamp: d.timestamp ?? "",
    });
  }
}

export class RFQSession {
  rfqId: string;
  request: MatchRequest;
  quotes: Quote[];
  selectedQuote: Quote | null;
  createdAt: string;

  constructor(request: MatchRequest) {
    this.rfqId = genId("rfq");
    this.request = request;
    this.quotes = [];
    this.selectedQuote = null;
    this.createdAt = nowIso();
  }

  addQuote(quote: Quote): void {
    this.quotes.push(quote);
  }

  rankQuotes(
    trustScores?: Record<string, number>,
    qualityWeight = 0.5,
    priceWeight = 0.5,
  ): Quote[] {
    const ts = trustScores ?? {};
    if (this.quotes.length === 0) return [];

    const maxAmount = Math.max(...this.quotes.map((q) => q.amount)) || 1.0;

    const scored: Array<[number, Quote]> = this.quotes.map((q) => {
      const priceScore = maxAmount > 0 ? (1.0 - q.amount / maxAmount) * 100 : 50;
      const trust = ts[q.agentId] ?? 50;
      const combined = priceWeight * priceScore + qualityWeight * trust;
      return [combined, q];
    });

    scored.sort((a, b) => b[0] - a[0]);
    return scored.map(([, q]) => q);
  }

  select(quoteId: string): Quote | null {
    for (const q of this.quotes) {
      if (q.quoteId === quoteId) {
        this.selectedQuote = q;
        return q;
      }
    }
    return null;
  }

  toDict(): Record<string, unknown> {
    return {
      rfq_id: this.rfqId,
      request_id: this.request.requestId,
      quotes: this.quotes.map((q) => q.toDict()),
      selected_quote: this.selectedQuote ? this.selectedQuote.toDict() : null,
      created_at: this.createdAt,
    };
  }
}

// ---------------------------------------------------------------------------
// Mechanism 3: Auction — Vickrey sealed-bid second-price (Section 6.3/8.2)
// ---------------------------------------------------------------------------

export interface BidDict {
  bid_id: string;
  agent_id: string;
  amount: number;
  currency: string;
  timestamp: string;
}

export class Bid {
  bidId: string;
  agentId: string;
  amount: number;
  currency: string;
  timestamp: string;

  constructor(opts: {
    bidId?: string;
    agentId?: string;
    amount?: number;
    currency?: string;
    timestamp?: string;
  } = {}) {
    this.bidId = opts.bidId || genId("bid");
    this.agentId = opts.agentId ?? "";
    this.amount = opts.amount ?? 0;
    this.currency = opts.currency ?? "USD";
    this.timestamp = opts.timestamp || nowIso();
  }

  toDict(): BidDict {
    return {
      bid_id: this.bidId,
      agent_id: this.agentId,
      amount: Math.round(this.amount * 10000) / 10000,
      currency: this.currency,
      timestamp: this.timestamp,
    };
  }

  static fromDict(d: Partial<BidDict>): Bid {
    return new Bid({
      bidId: d.bid_id ?? "",
      agentId: d.agent_id ?? "",
      amount: d.amount ?? 0,
      currency: d.currency ?? "USD",
      timestamp: d.timestamp ?? "",
    });
  }
}

export interface AuctionResultDict {
  auction_id: string;
  format: string;
  winner_id: string;
  winning_bid: number;
  clearing_price: number;
  total_bids: number;
  timestamp: string;
}

export class AuctionResult {
  auctionId: string;
  format: string;
  winnerId: string;
  winningBid: number;
  clearingPrice: number;
  totalBids: number;
  timestamp: string;

  constructor(opts: {
    auctionId?: string;
    auctionFormat?: string;
    winnerId?: string;
    winningBid?: number;
    clearingPrice?: number;
    totalBids?: number;
    timestamp?: string;
  } = {}) {
    this.auctionId = opts.auctionId || genId("auction");
    this.format = opts.auctionFormat ?? "vickrey";
    this.winnerId = opts.winnerId ?? "";
    this.winningBid = opts.winningBid ?? 0;
    this.clearingPrice = opts.clearingPrice ?? 0;
    this.totalBids = opts.totalBids ?? 0;
    this.timestamp = opts.timestamp || nowIso();
  }

  toDict(): AuctionResultDict {
    return {
      auction_id: this.auctionId,
      format: this.format,
      winner_id: this.winnerId,
      winning_bid: Math.round(this.winningBid * 10000) / 10000,
      clearing_price: Math.round(this.clearingPrice * 10000) / 10000,
      total_bids: this.totalBids,
      timestamp: this.timestamp,
    };
  }
}

export function vickreyAuction(
  bids: Bid[],
  minTrustScore = 0,
  trustScores?: Record<string, number>,
): AuctionResult | null {
  const ts = trustScores ?? {};

  const qualified = bids.filter(
    (b) => (ts[b.agentId] ?? 100) >= minTrustScore,
  );

  if (qualified.length === 0) return null;

  qualified.sort((a, b) => a.amount - b.amount);

  const winner = qualified[0];
  const clearingPrice = qualified.length > 1 ? qualified[1].amount : winner.amount;

  return new AuctionResult({
    auctionFormat: "vickrey",
    winnerId: winner.agentId,
    winningBid: winner.amount,
    clearingPrice,
    totalBids: bids.length,
  });
}

export function englishAuction(
  bids: Bid[],
  minTrustScore = 0,
  trustScores?: Record<string, number>,
): AuctionResult | null {
  const ts = trustScores ?? {};

  const qualified = bids.filter(
    (b) => (ts[b.agentId] ?? 100) >= minTrustScore,
  );

  if (qualified.length === 0) return null;

  qualified.sort((a, b) => a.amount - b.amount);
  const winner = qualified[0];

  return new AuctionResult({
    auctionFormat: "english",
    winnerId: winner.agentId,
    winningBid: winner.amount,
    clearingPrice: winner.amount,
    totalBids: bids.length,
  });
}
